### PR TITLE
Feat/fingerprint

### PR DESCRIPTION
**Pull request by caleb-bitaccess**
# Description of the change

add deriveUniPubKeyForPath and deriveUniPubKeyForPath to all coin's paymentsUtils

### CODE REVIEW

File: packages/coinlib-bitcoin-cash/src/helpers.ts [hunk 1]
```diff
@@ -158,3 +166,57 @@ export function estimateBitcoinCashTxSize(
...
+export function deriveUniPubKeyForPath(seed: Buffer, derivationPath: string): string {
+  const splitPath = derivationPath.split('/')
+  if (splitPath?.length !== 4 || splitPath[0] !== 'm') {
+    throw new TypeError(`Invalid derivationPath ${derivationPath}`)
+  }
+
+  const bitcoincashSupportedPurposes = [`44'`, `87'`]
+  const purpose = splitPath[1]
+  if (!bitcoincashSupportedPurposes.includes(purpose)) {
+    throw new TypeError(`Purpose in derivationPath ${purpose} not supported by Bitcoincash`)
```
Comment (dylanseago) [inline packages/coinlib-bitcoin-cash/src/helpers.ts:206]: No need to enforce this. You can use any custom derivation path you like with any address type you like. The purposes are just a convention. There could very likely be a customer that wants to use `45'` or `48'` as they're common multisig paths, or something else entirely custom.

File: packages/coinlib-bitcoin-cash/src/helpers.ts [hunk 2]
```diff
@@ -158,3 +166,57 @@ export function estimateBitcoinCashTxSize(
...
+  }
+
+  const coinType = splitPath[2]
+  let network: BitcoinjsNetwork | null = null
+  if (coinType === `${BITCOINCASH_COINTYPE_MAINNET}'`) {
+    network = NETWORK_MAINNET
+  } else if (coinType === `${BITCOINCASH_COINTYPE_TESTNET}'`) {
+    network = NETWORK_TESTNET
+  } else {
+    throw new TypeError(`Invalid derivationPath coin type ${coinType}`)
```
Comment (dylanseago) [inline packages/coinlib-bitcoin-cash/src/helpers.ts:216]: In order to avoid this whole if/else structure you could just require `networkType` as a parameter. This would reduce the edge cases and be totally reasonable because the utils calling this helper already have `networkType` as an instance variable.

File: packages/coinlib-bitcoin-cash/test/BitcoinCashPaymentsUtils.test.ts [hunk 3]
```diff
@@ -101,4 +103,55 @@ describe('BitcoinCashPaymentUtils', () => {
...
+      expect(functionToTrow).toThrow(`Purpose in derivationPath 84' not supported by Bitcoincash`)
+    })
+
+    test('Testnet MultisigLegacy', () => {
+      const xpub = puMainnet.deriveUniPubKeyForPath(seedBuffer, `m/87'/1'/4'`)
+      const expectedXpub =
+        'tpubDCHyWQfXp2aefcHocJn7CvtRJ1mmqbm2x8D1v3HJ8QF9cDoBK25327B42W5HtyvEmfVZB3f3zVHWwQp4BcWKosQSuugqCDZH7KFvmtbtSwz'
+      expect(xpub).toBe(expectedXpub)
+    })
+  })
```
Comment (dylanseago) [inline packages/coinlib-bitcoin-cash/test/BitcoinCashPaymentsUtils.test.ts:156]: Great tests!

File: packages/coinlib-bitcoin/src/bitcoinish/constants.ts [hunk 4]
```diff
@@ -1,8 +1,25 @@
...
   [FeeLevel.High]: 1,
   [FeeLevel.Medium]: 24,
   [FeeLevel.Low]: 144,
 }
+
+export const BITCOIN_COINTYPE_MAINNET =  `0`
+export const BITCOIN_COINTYPE_TESTNET =  `1`
+
+export const NETWORK_MAINNET = networks.bitcoin
+export const NETWORK_TESTNET = networks.testnet
```
Comment (dylanseago) [inline packages/coinlib-bitcoin/src/bitcoinish/constants.ts:16]: I believe these 4 constants above would be more appropriate in `src/constants.ts` since they're bitcoin specific and not generally applicable to all bitcoinish coins

File: packages/coinlib-bitcoin/src/helpers.ts [hunk 5]
```diff
@@ -62,3 +76,48 @@ export function estimateBitcoinTxSize(
...
+    network = NETWORK_MAINNET
+  } else if (coinType === `${BITCOIN_COINTYPE_TESTNET}'`) {
+    network = NETWORK_TESTNET
+  } else {
+    throw new TypeError(`Invalid derivationPath coin type ${coinType}`)
+  }
+
+  const root = bip32.fromSeed(seed, network)
+  const account = root.derivePath(derivationPath)
+  return account.neutered().toBase58()
```
Comment (dylanseago) [inline packages/coinlib-bitcoin/src/helpers.ts:122]: Could both of these helpers be generalized and moved into bitcoinish or common instead? I see a lot of repetitive logic that could be deduplicated.

Something like:
```
function createDeterminePathForIndexHelper(constants: {
  addressTypePurposes: { [addressType: string]: string },
  coinTypes: { [networkType in NetworkType]: string },
}) {
  return (
    accountIndex: number,
    addressType?: BitcoinishAddressType,
    networkType?: NetworkType,
  ): string => {
    // implementation goes here and uses `constants` param provided through closure 
  }
}
```

A similar pattern can be adopted for `deriveUniPubKeyForPath`

File: packages/coinlib-common/src/PaymentsUtils.ts [hunk 6]
```diff
@@ -107,4 +108,12 @@ export interface PaymentsUtils {
   ): Promise<BaseTransactionInfo>
 
   getBlock(id?: string | number): Promise<BlockInfo>
+
+  determinePathForIndex(
+    accountIndex: number,
+    addressType?: BitcoinishAddressType,
```
Comment (dylanseago) [inline packages/coinlib-common/src/PaymentsUtils.ts:114]: Moving this `BitcoinishAddressType` into common isn't really appropriate since it's only applicable to bitcoinish coins, not all networks generally. I'd recommend following the pattern of the other methods above that also have optional parameters that may or may not be used in different ways by different networks.

For example:
```
isValidAddress<O extends { format?: string }>(address: string, options?: O): boolean
```
The `format` field here is an enum for ethereum style networks, but unused by other networks. You could do something similar for addressType so that it's general enough to be used for both bitcoinish address types, and any other potential variability in other networks addresses.

```
determinePathForIndex<O extends { addressType?: string }>(accountIndex: number, options?: O): string
```

Then inside your bitcoinish implementations you'd just use:
```
const addressType = options?.addressType ? assertType(AddressTypeT, options?.addressType) : DEFAULT_ADDRESS_TYPE
```

**Review by dylanseago — CHANGES_REQUESTED** [review]

**Comment by dylanseago** [comment]
@caleb-bitaccess Have all the previous review comments been addressed? If so please mark them as resolved

File: packages/coinlib-bitcoin-cash/src/helpers.ts (hunk 1 above)
Comment (caleb-bitaccess) [reply packages/coinlib-bitcoin-cash/src/helpers.ts]: restriction removed

File: packages/coinlib-bitcoin/src/bitcoinish/constants.ts (hunk 4 above)
Comment (caleb-bitaccess) [reply packages/coinlib-bitcoin/src/bitcoinish/constants.ts]: removed

File: packages/coinlib-bitcoin/src/helpers.ts (hunk 5 above)
Comment (caleb-bitaccess) [reply packages/coinlib-bitcoin/src/helpers.ts]: put them in bitcoinish helper

File: packages/coinlib-common/src/PaymentsUtils.ts (hunk 6 above)
Comment (caleb-bitaccess) [reply packages/coinlib-common/src/PaymentsUtils.ts]: removed BitcoinishAddressType from coinlib-common

File: packages/coinlib-tron/src/helpers.ts [hunk 7]
```diff
@@ -53,3 +53,40 @@ export function privateKeyToAddress(privateKey: string): string {
     throw new Error('Validation failed for address derived from private key')
   }
 }
+
+export function determinePathForIndex(accountIndex: number, addressType?: any, networkType?: NetworkType): string {
+  const LEGENCY = 'p2pkh'
+  if (addressType && addressType?.toString() !== LEGENCY) {
+    throw new TypeError(`Tron does not support this type ${addressType}`)
+  }
```
[the remark this answers is not in the document]
Comment (caleb-bitaccess) [reply packages/coinlib-tron/src/helpers.ts]: checking removed

File: packages/coinlib/src/utils.ts [hunk 8]
```diff
@@ -1,4 +1,18 @@
...
+export function getSeedHexFromMnemonic(mnemonic: string): string {
+  const seed = bip39.mnemonicToSeedSync(mnemonic)
+  return seed.toString("hex")
+}
+
+export function getFingerprintFromSeedHex(seedHex: string): string {
+  const seed = Buffer.from(seedHex, 'hex')
+  const root = bip32.fromSeed(seed)
+  return root.fingerprint.toString('hex')
+}
```
[the remark this answers is not in the document]
Comment (caleb-bitaccess) [reply packages/coinlib/src/utils.ts:18]: added getFingerprint() to  CoinPayments, but I kept these two methods here.

File: packages/coinlib-ethereum/src/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import { NetworkType, bip32 } from '@bitaccess/coinlib-common'
+import { ETHEREUM_COINTYPE_MAINNET, ETHEREUM_COINTYPE_TESTNET } from './constants'
+
+export function determinePathForIndex(accountIndex: number, addressType?: any, networkType?: NetworkType): string {
+  const LEGENCY = 'p2pkh'
+  if (addressType && addressType?.toString() !== LEGENCY) {
```
[the remark this answers is not in the document]
Comment (caleb-bitaccess) [reply packages/coinlib-ethereum/src/helpers.ts]: removed

**Review by dylanseago — APPROVED** [review]
